adds http-checksum trait support (#741)
awslabs/smithy: docs/source/1.0/spec/core/behavior-traits.rst

            input: PutSomethingInput,
            output: PutSomethingOutput
        }

.. _httpChecksum-trait:

----------------------
``httpChecksum`` trait
----------------------

Summary
    Indicates that an operation's HTTP Request or Response supports checksum
    validation. At least one of request or response checksum properties must
    be specified within the trait.

>     When this trait is modeled along with

Reviewer: This trait needs a section to define behavior for it [like `@paginated`](https://awslabs.github.io/smithy/1.0/spec/core/behavior-traits.html#paginated-trait), instead of sprinkling it in to the trait documentation.

Author: Defined a behavior section.

        The recommended ABNF to follow for prefix is `([a-z]*([a-z]/[0-9]/"-")*`.

        To construct a header name for usage with checksum, follow the pattern:
        `prefix + algorithm-name`. For example, if prefix is "x-checksum-"
        and "sha256" algorithm is used for computing checksum, header name
        would resolve as "x-checksum-sha256".

    * - location
      - ``string``
      - A string representing checksum location. The string value MUST be a
        valid :ref:`Location enum <checksum-location-enum>`. By default,
        location resolves to "header" value. Service MUST always support
        checksum in header for both requests and responses.

        Optionally, for request, location can be modeled as "trailer"
        indicating service supports checksum in trailer, in addition to
        support in header.

    * - algorithms
      - ``[string]``
      - **Required**. List of strings representing checksum algorithms
        supported for the HTTP Request or Response. Algorithms is a
        required member, and a valid algorithm name SHOULD follow
        ABNF: `([a-z]* [0-9]*)*`. Algorithm name SHOULD be lower-cased
        and not hyphenated.

> .. tabs::

Reviewer: Examples for traits are usually under the documentation for the trait itself, so this should move up.

    .. code-tab:: smithy

        @httpChecksum(
            request: {
                location: "trailer",
                prefix: "x-checksum-",
                algorithms: ["sha256", "crc32"]
            },
            response: {
                prefix: "x-checksum-",
                algorithms: ["sha256", "crc32"]
            }
        )
        operation PutSomething {
            input: PutSomethingInput,
            output: PutSomethingOutput
        }

> .. _checksum-location-enum:

Reviewer: This is small enough where it doesn't need its own section, it can just be part of the property explanation (like [the `error` trait value type](https://awslabs.github.io/smithy/1.0/spec/core/type-refinement-traits.html#error-trait).)

                            }
                        }
                    }
                }
            }
        }

.. _checksum-properties:

HttpChecksumProperties structure
================================

HttpChecksumProperties defines checksum validation behavior using the
following members:

.. list-table::
    :header-rows: 1
    :widths: 10 10 80

    * - Property
      - Type
      - Description
    * - prefix
      - ``string``
      - **Required**. The prefix string is used to construct a header for a
        checksum type. Prefix is a required member, and SHOULD be lower-cased.
        The recommended ABNF to follow for prefix is `([a-z]*([a-z]/[0-9]/"-")*`.

        To construct a header name for usage with checksum, follow the pattern:
        `prefix + algorithm-name`. For example, if prefix is "x-checksum-"
        and "sha256" algorithm is used for computing checksum, header name
        would resolve as "x-checksum-sha256".

    * - location
      - ``string``
      - A string representing location where checksum is supplied. The string
        value MUST be set to "header" or "trailer". By default, location

>         resolves to "header" value. Service MUST always support checksum in

Author: Updated to use a list of locations - all supported locations will now be explicit.

        header for both requests and responses.

        Optionally, for request, location can be modeled as "trailer"
        indicating service supports checksum in trailer, in addition to
        support in header.

    * - algorithms
      - ``[string]``
      - **Required**. List of strings representing checksum algorithms
        supported for the HTTP Request or Response. Algorithms is a
        required member, and a valid algorithm name SHOULD follow
        ABNF: `([a-z]* [0-9]*)*`. Algorithm name SHOULD be lower-cased
        and not hyphenated.

.. _checksum_trait with_checksum_required:

> HttpChecksumRequired Behavior

Author: Was not sure what other behavior needs to be called out in this subsection. I added statement around client and server behavior as suggested.

        or trailer name will be "x-checksum-sha256".

        Recommended ABNF for prefix is as follows:

        .. code-block:: abnf

            prefix = lower-alpha ["-"] *((lower-alpha / DIGIT)["-"])
            lower-alpha = %x61-7A

    * - locations
      - ``[string]``
      - A priority list representing supported locations where checksum can be
        supplied. Valid members for locations list are "header", and "trailer".
        If not provided, locations resolve to a list containing "header" value,
        indicating "header" location is supported to supply checksum.

    * - algorithms
      - ``[string]``
      - **Required**. List of string representing checksum algorithms
        supported for the HTTP Request or Response. A valid algorithm
        should follow ABNF:

        .. code-block:: abnf

            algorithms = 1*(lower-alpha / DIGIT)
            lower-alpha = %x61-7A

.. _checksum_trait with_checksum_required:

HttpChecksumRequired Behavior
=============================

When `httpChecksum` trait is modeled along with
:ref:`httpChecksumRequired trait <httpChecksumRequired-trait>`,

> service MUST accept a checksum value sent as per the `httpChecksum` trait's

Reviewer 2: s/service/the service/g

            }
        }

.. _checksum-properties:

HttpChecksumProperties structure
================================

``HttpChecksumProperties`` defines checksum validation behavior using the
following members:

.. list-table::
    :header-rows: 1
    :widths: 10 10 80

    * - Property
      - Type
      - Description
    * - prefix
      - ``string``
      - **Required**. The prefix string is used to construct a header or
        trailer name for a checksum type. To construct a header or trailer
        name for usage with checksum, follow the pattern:
        ``prefix + algorithm-name``.

        For example, if prefix is ``x-checksum-``
        and ``sha256`` algorithm is used for computing checksum, resolved header
        or trailer name will be ``x-checksum-sha256``.

        Recommended ABNF for prefix is as follows:

        .. code-block:: abnf

            prefix = lower-alpha *(["-"](lower-alpha / DIGIT)) "-"
            lower-alpha = %x61-7A

        That is, it should start with a lowercase letter followed by any
        number of lowercase letters, digits, or non-sequential hyphens,
        and it should end in a hyphen.

    * - locations
      - ``[string]``

>       - A priority-ordered list representing supported locations where checksum
>         can be supplied. Valid values are ``header``, and ``trailer``.
>         The default value is a list only containing ``header``.

Reviewer 2: I know I contributed to this problem, but:

```suggestion
      - A priority-ordered list representing supported locations where checksum
        can be supplied. Valid values are:
        
        * ``header`` - Indicates the checksum should be placed in an HTTP header.
        * ``trailer`` - Indicates the checksum should be placed in the `chunked trailer part`_
          of the body.
        
        The default value is a list only containing ``header``.
```

    * - algorithms
      - ``[string]``
      - **Required**. List of string representing checksum algorithms
        supported for the HTTP request or response. A valid algorithm
        should follow ABNF:

        .. code-block:: abnf

            algorithms = 1*(lower-alpha / DIGIT)
            lower-alpha = %x61-7A

        That is, it should be comprised only of digits and lowercase letters.

.. _httpChecksum_trait_behavior:

HttpChecksum Behavior
=====================

.. rubric:: Client Behavior

> If the ``httpChecksum`` trait has a modeled ``request`` section, for HTTP request,
> the client MUST compute the request payload checksum, using an algorithm
> defined in the algorithms property. This computed checksum MUST be supplied
> at the first supported location defined in the locations property of the trait.
> Locations property supports modeling ``header``, and ``trailer`` as locations.
> The header or trailer name MUST be constructed using the prefix property as
> suggested in the :ref:`HTTP checksum properties<checksum-properties>` section.

Reviewer 2: ```suggestion
If the ``httpChecksum`` trait has a modeled ``request`` section, for HTTP request,
the client MUST compute the request payload checksum, using an algorithm
defined in the algorithms property. This computed checksum MUST be supplied
at one of the locations defined in the location property of the trait. It SHOULD be
supplied at the first supported location defined in the locations property of the trait.
If the resolved location is ``header``, the client MUST put the checksum into the
HTTP request headers. If the resolved location is ``trailer``, the client MUST put
the checksum into the `chunked trailer part`_ of the body. The header or trailer
field name MUST be constructed using the prefix property as described in the
:ref:`HTTP checksum properties<checksum-properties>` section.
```

Then at the bottom you'll need:

```rst
.. _chunked trailer part: https://tools.ietf.org/html/rfc7230#section-4.1.2
```

Author: Updated

> If the ``httpChecksum`` trait has a modeled ``response`` section, for HTTP
> response, the client MUST look for checksum at supported locations as per
> defined properties. If a checksum is found, the client MUST validate the received
> checksum value by computing corresponding checksum of the received payload.

Reviewer 2: ```suggestion
If the ``httpChecksum`` trait has a modeled ``response`` section, for HTTP
response, the client MUST look for a checksum at supported locations as per
defined properties. If a checksum is found, the client MUST validate the received
checksum value by computing the corresponding checksum of the received payload.
```

.. rubric:: Service Behavior

If the ``httpChecksum`` trait has a modeled ``request`` section, for HTTP request,

> the service SHOULD validate the received checksum, by computing corresponding

Reviewer 2: ```suggestion
the service SHOULD validate the received checksum by computing corresponding
```

checksum of the request payload to ensure data integrity.

> If the ``httpChecksum`` trait models a ``response`` section, for HTTP response,
> the service SHOULD send at least one supported payload checksums at the first
> supported location in locations property. The header or trailer name used,
> MUST be constructed using modeled prefix property as suggested in the
> :ref:`HTTP checksum properties<checksum-properties>` section.

Reviewer 2: ```suggestion
If the ``httpChecksum`` trait models a ``response`` section, for HTTP response,
the service SHOULD send at least one supported payload checksum at the first
supported location in the locations property. The header or trailer name used,
MUST be constructed using modeled prefix property as suggested in the
:ref:`HTTP checksum properties<checksum-properties>` section.
```

            input: PutSomethingInput,
            output: PutSomethingOutput
        }

.. _httpChecksum-trait:

----------------------
``httpChecksum`` trait
----------------------

Summary
    Indicates that an operation's HTTP request or response supports checksum
    validation. At least one of request or response checksum properties must
    be specified within the trait.

Trait selector
    ``operation``
Value type
    ``structure``

The ``httpChecksum`` trait is a structure that contains the following members:

.. list-table::
    :header-rows: 1
    :widths: 10 10 80

    * - Property
      - Type
      - Description
    * - request
      - :ref:`HttpChecksumProperties structure <checksum-properties>`
      - The ``request`` property defines checksum validation behavior for
        HTTP requests.

    * - response
      - :ref:`HttpChecksumProperties structure <checksum-properties>`
      - The ``response`` property defines checksum validation behavior for
        HTTP responses.

.. tabs::

    .. code-tab:: smithy

>         @httpChecksum(

Reviewer 3: Did we consider making these a little more explicit, customizable, and future proof by using a map similar to:

```
@httpChecksum(
    request: {
        sha256: {
            header: "x-checksum-sha256"
        },
        crc32: {
            header: "x-checksum-crc32"
        }
    }
)
```

Here, `request` and `response` are maps where each key is the algorithm, and each value is a union of possible locations (i.e., it can be set to `header` or `trailer`, and we could expand to `query` in the future for requests if needed). The value of the union is the binding location name (i.e., header name or trailing header name).

This provides more flexibility than the current proposal because each algorithm can use whatever header or query string parameter is necessary. The header/query binding is explicit and does not need prefixes. The names can now support anything, including `Content-MD5` if we ever wanted to support it (we probably don't but just an example).

Implementations don't need to perform any prefix logic to determine the header name either.

Author: I like the idea, have few concerns. Let's sync offline on this.

HttpChecksum Behavior
=====================

.. rubric:: Client Behavior

If the ``httpChecksum`` trait has a modeled ``request`` section, for HTTP request,
the client MUST compute the request payload checksum, using an algorithm
defined in the algorithms property. This computed checksum MUST be supplied
at one of the locations defined in the location property of the trait. It SHOULD be
supplied at the first supported location defined in the locations property of the trait.
If the resolved location is ``header``, the client MUST put the checksum into the
HTTP request headers. If the resolved location is ``trailer``, the client MUST put
the checksum into the `chunked trailer part`_ of the body. The header or trailer
field name MUST be constructed using the prefix property as described in the
:ref:`HTTP checksum properties<checksum-properties>` section.

If the ``httpChecksum`` trait has a modeled ``response`` section, for HTTP
response, the client MUST look for a checksum at supported locations as per
defined properties. If a checksum is found, the client MUST validate the received
checksum value by computing the corresponding checksum of the received payload.

> 

Reviewer: Can add a label here to be referable to from above.

```
::  _httpChecksum_trait_header_conflict_behavior:
```

> Service MAY model input or output shape member bound to an HTTP header

Reviewer: An operation's `input`, `output`, and `error` structures MAY contain a member bound...

            input: PutSomethingInput,
            output: PutSomethingOutput
        }

.. _httpChecksum-trait:

----------------------
``httpChecksum`` trait
----------------------

Summary
    Indicates that an operation's HTTP Request or Response supports checksum
    validation. At least one of request or response checksum properties must
    be specified within the trait.
Trait selector
    ``operation``
Value type
    ``structure``

The `httpChecksum` trait is a structure that contains the following members:

.. list-table::
    :header-rows: 1
    :widths: 10 10 80

    * - Property
      - Type
      - Description

>     * - Request

Reviewer 3: Can you make these lowercase?

Author: updated `Request` -> `request`, and `Response` -> `response`.

      - ``[`` :ref:`HttpChecksumProperties structure <checksum-properties>` ``]``
      - Request structure values define checksum validation behavior for
        HTTP Request.

    * - Response
      - ``[`` :ref:`HttpChecksumProperties structure <checksum-properties>` ``]``
      - Response structure values define checksum validation behavior for

>         HTTP Response.

Author: Added a blurb at top

.. - checksum-properties:

HTTPChecksumProperties structure
================================

HTTPChecksumProperties defines checksum validation behavior using the
following members:

.. list-table::
    :header-rows: 1
    :widths: 10 10 80

    * - Property
      - Type
      - Description
    * - prefix
      - ``string``

>       - **Required**. The prefix string is used to construct a header for a

Author: Good call out, `prefix`+`algorithm` will be used. I thought of this, wasn't sure if we should document it as a constraint/recommendation.

But I think it makes sense to have it as a recommendation, and add protocol tests around it.

>       - **Required**. The prefix string is used to construct a header for a

Reviewer 4: :+1: I think the clarification in the docs makes this explicit enough.

        checksum type. Prefix is a required member, and SHOULD be lower-cased. The
        recommended ABNF to follow for prefix is `([a-z]*([a-z]/[0-9]/"-")*`.

>     * - location

Reviewer: Is there a reason this property isn't an enum when the valid locations are "header" and "trailer"?

Author: No reason, good call out, will define a LOCATION enum.
